kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/SchemaUtils.java

```diff
@@ -79,6 +83,356 @@ public static void validateSchema(StructType schema, boolean isColumnMappingEnab
     validateSupportedType(schema);
   }
 
+  /**
+   * Performs the following validations on an updated table schema using the current schema as a
+   * base for validation. ColumnMapping must be enabled to call this
+   *
+   * <p>The following checks are performed:
+   *
+   * <ul>
+   *   <li>No duplicate columns are allowed
+   *   <li>Column names contain only valid characters
+   *   <li>Data types are supported
+   *   <li>No new non-nullable fields are added or no tightening of nullable fields
+   *   <li>Physical column name consistency is preserved in the new schema
+   *   <li>Partition columns are not modified
+   *   <li>Nested IDs for array/map types are preserved in the new schema
+   *   <li>No type changes
+   * </ul>
+   *
+   * @return SchemaChanges which encapsulate the additions/removals/updates
+   */
+  public static SchemaChanges validateUpdatedSchema(
+      StructType currentSchema,
+      StructType newSchema,
+      Set<String> expectedPartitionColumns,
+      Metadata metadata) {
+    checkArgument(
+        isColumnMappingModeEnabled(
+            ColumnMapping.getColumnMappingMode(metadata.getConfiguration())));
+    validateSchema(newSchema, true /*columnMappingEnabled*/);
+    validatePartitionColumns(newSchema, new ArrayList<>(expectedPartitionColumns));
+    return validateSchemaEvolution(
+        currentSchema,
+        newSchema,
+        ColumnMapping.getColumnMappingMode(metadata.getConfiguration()),
+        metadata);
+  }
+
+  /* Validate if a given schema evolution is safe for a given column mapping mode */
+  private static SchemaChanges validateSchemaEvolution(
+      StructType currentSchema,
+      StructType newSchema,
+      ColumnMappingMode columnMappingMode,
+      Metadata metadata) {
+    switch (columnMappingMode) {
+      case ID:
+      case NAME:
+        return validateSchemaEvolutionById(currentSchema, newSchema, metadata);
```

Comment:
Down the line we can add something like 


```
abstract class SchemaEvolutionValidator {
    abstract validate(StructType schemaBefore, StructType schemaAfter);
}

class SchemaEvolutionValidatorById {
     validate(StructType schemaBefore, StructType schemaAfter) {
          SchemaChanges changes = computeDiffById(schemaBefore, schemaAfter); 
           doAll the validation
      }
}

```

Then we can move all this logic out of utils, and later if required have an implementation for NONE based schema evolution which would allow evolution like column additions.

Reply:
But I didn't want to possibly overengineer in this PR